context_test.go

```diff
@@ -431,6 +436,165 @@ func TestUnsupportedSubProtocol(t *testing.T) {
 	}
 }
 
+func TestHandshake(t *testing.T) {
+	serverCtx, err := NewContext(ContextOptions{ProtocolIds: []string{"ServerProtocol"}})
+	require.NoError(t, err)
+	serverCtx.LogMessages = true
+	serverCtx.LogFrames = true
+
+	server := serverCtx.WebSocketServer()
+	wg := sync.WaitGroup{}
+	wg.Add(1)
+	server.PostHandshakeCallback = func(err error) {
+		defer wg.Done()
+		require.Error(t, err)
+	}
+	mux := http.NewServeMux()
+	mux.Handle("/someBlip", server)
+	listener, err := net.Listen("tcp", ":0")
+	require.NoError(t, err)
+	defer func() {
+		assert.NoError(t, listener.Close())
+	}()
+
+	go func() {
+		// error will happen when listener closes in defer
+		_ = http.Serve(listener, mux)
+	}()
+
+	// Client
+	client, err := NewContext(ContextOptions{ProtocolIds: []string{"ClientProtocol"}})
+	require.NoError(t, err)
+
+	port := listener.Addr().(*net.TCPAddr).Port
+	destUrl := fmt.Sprintf("ws://localhost:%d/someBlip", port)
+
+	_, err = client.Dial(destUrl)
+	require.Error(t, err)
+	require.NoError(t, WaitWithTimeout(&wg, time.Second*10))
+}
+
+func TestOrigin(t *testing.T) {
+	protocol := "Protocol1"
+	testCases := []struct {
+		serverOrigin  []string
+		requestOrigin string
+		hasError      bool
+	}{
+		{
+			serverOrigin:  []string{"example.com"},
+			requestOrigin: "https://example.com",
+		},
+		{
+			serverOrigin:  []string{"foobar.com", "example.com"},
+			requestOrigin: "https://example.com",
+		},
+		{
+			serverOrigin:  []string{"*"},
+			requestOrigin: "https://example.com",
+		},
+		{
+			serverOrigin:  []string{"*"},
```

Comment:
Should we also test:
- Unset request Origin
  - Like CBL/non-browser clients do
```go
serverOrigin: []string{"example.com"},
requestOrigin: null,

...

if test.requestOrigin != nil {
	config.HTTPHeader = make(http.Header)
	config.HTTPHeader.Add("Origin", test.requestOrigin)
}
```
- Implicitly matching request origin
  - For webapps on the same domain as SG without explicit CORS settings declared)
```go
serverOrigin: null,
requestOrigin: "http://localhost",
```